src/swift/Dispatch.swift

```diff
@@ -11,45 +11,156 @@
 //===----------------------------------------------------------------------===//
 
 @_exported import Dispatch
-
-/// The type of blocks submitted to dispatch queues, which take no arguments
-/// and have no return value.
-///
-/// The dispatch_block_t typealias is different from usual closures in that it
-/// uses @convention(block). This is to avoid unnecessary bridging between
-/// C blocks and Swift closures, which interferes with Grand Central Dispatch
-/// APIs that depend on the referential identity of a block.
-public typealias dispatch_block_t = @convention(block) () -> Void
+import CDispatch
 
 //===----------------------------------------------------------------------===//
-// Macros
-// FIXME: rdar://16851050 update API so these import better
+// Linux-specific overlay layer to compensate for lack of Objective-C
+//
+//   CDispatch classes and APIs are wrapped in Swift objects/functions
+//   to achieve API compatibility with Darwin platforms and integration with
+//   Swift retain/release operations.
+//
+//   Injecting an extra level of wrapping is sub-optimal, but pushing the Swift
+//   object model down into the C libdispatch implementation is too invasive
+//   a change to undertake at this stage of development.
 //===----------------------------------------------------------------------===//
 
-// dispatch/io.h
-public var DISPATCH_IO_STREAM: dispatch_io_type_t {
-  return 0
+//////////
+// Define Swift-level class hierarchy for subset of dispatch types
+// returned from API methods with DISPATCH_RETURNS_RETAINED
+//////////
+
+public typealias dispatch_object_t = DispatchObject
+public class DispatchObject {
```

Comment:
I think either could work.  I was thinking it might be useful to distinguish between the native and Swift layers of the code, but I really didn't use the capitalized class names outside the declarations.  I could rework if you want.